File: backend/app/schemas/sentence_builder.py
"""
Sentence Builder Schemas.

Pydantic models for Duolingo-style sentence building activities.
Students arrange jumbled words into correct sentence order.
"""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


class SentenceBuilderRequest(BaseModel):
    """
    Request to generate a sentence builder activity.

    Attributes:
        book_id: ID of the book to extract sentences from.
        module_ids: Optional list of specific module IDs. If None, uses all modules.
        sentence_count: Number of sentences in the activity (1-10).
        difficulty: Difficulty level based on sentence length.
        include_audio: Whether to include audio for correct sentences.
    """

    book_id: int
    module_ids: list[int] | None = Field(
        default=None,
        description="Specific modules to use. If None, uses all modules.",
    )
    sentence_count: int = Field(
        default=10,
        ge=1,
        le=50,
        description="Number of sentences (1-50).",
    )
    difficulty: Literal["easy", "medium", "hard"] = Field(
        default="medium",
        description="Difficulty: easy (4-6 words), medium (7-10 words), hard (11+ words).",
    )
    include_audio: bool = Field(
        default=True,
        description="Whether to include TTS audio for correct sentences.",
    )


class SentenceBuilderItem(BaseModel):
    """
    A single sentence item in the activity.

    Contains the correct sentence and shuffled word bank.

    Attributes:
        item_id: Unique identifier for this sentence.
        correct_sentence: The correct sentence as a string.
        words: Shuffled word bank for the student to arrange.
        word_count: Number of words in the sentence.
        audio_url: TTS audio URL for the correct sentence.
        source_module_id: ID of the module this sentence came from.
        source_page: Page number in the module (if available).
        difficulty: Difficulty level of this sentence.
    """

    item_id: str
    correct_sentence: str
    words: list[str]  # Shuffled word bank
    word_count: int
    audio_url: str | None = None
    source_module_id: int
    source_page: int | None = None
    difficulty: str


class SentenceBuilderActivity(BaseModel):
    """
    A complete sentence builder activity.

    This is the full activity representation stored internally,
    including the correct sentences.

    Attributes:
        activity_id: Unique identifier for the activity.
        book_id: ID of the book the activity is generated from.
        module_ids: List of module IDs used for the activity.
        sentences: List of sentence items with correct answers.
        difficulty: Difficulty level of the activity.
        include_audio: Whether audio is included.
        created_at: When the activity was generated.
    """

    activity_id: str
    book_id: int
    module_ids: list[int]
    sentences: list[SentenceBuilderItem]
    difficulty: str
    include_audio: bool
    created_at: datetime


class SentenceBuilderActivityPublic(BaseModel):
    """
    Public version of activity without correct sentence strings.

    Returned to students when they access an activity for taking.
    The correct_sentence field is omitted from items.

    Attributes:
        activity_id: Unique identifier for the activity.
        book_id: ID of the book the activity is generated from.
        module_ids: List of module IDs used for the activity.
        sentences: List of sentence items (words only, no correct sentence).
        difficulty: Difficulty level of the activity.
        include_audio: Whether audio is included.
        created_at: When the activity was generated.
        sentence_count: Total number of sentences.
    """

    activity_id: str
    book_id: int
    module_ids: list[int]
    sentences: list["SentenceBuilderItemPublic"]
    difficulty: str
    include_audio: bool
    created_at: datetime
    sentence_count: int


class SentenceBuilderItemPublic(BaseModel):
    """
    Public version of a sentence item without the correct answer.

    Attributes:
        item_id: Unique identifier for this sentence.
        words: Shuffled word bank for the student to arrange.
        word_count: Number of words in the sentence.
        difficulty: Difficulty level of this sentence.
    """

    item_id: str
    words: list[str]
    word_count: int
    difficulty: str


class SentenceBuilderSubmission(BaseModel):
    """
    Student's submission of sentence answers.

    Maps sentence item IDs to the student's word ordering.

    Attributes:
        answers: Dictionary mapping item_id to ordered list of words.
    """

    answers: dict[str, list[str]] = Field(
        description="Map of item_id to ordered list of words.",
    )


class SentenceResult(BaseModel):
    """
    Result for a single sentence after submission.

    Attributes:
        item_id: ID of the sentence item.
        submitted_words: The student's submitted word order.
        correct_sentence: The correct sentence.
        is_correct: Whether the submission was correct.
        audio_url: Audio URL for the correct sentence (if available).
    """

    item_id: str
    submitted_words: list[str]
    correct_sentence: str
    is_correct: bool
    audio_url: str | None = None


class SentenceBuilderResult(BaseModel):
    """
    Complete results of a submitted sentence builder activity.

    Returned after a student submits their answers, revealing
    the correct sentences and calculating the score.

    Attributes:
        activity_id: ID of the activity.
        student_id: ID of the student who submitted.
        score: Number of correct sentences.
        total: Total number of sentences.
        percentage: Score as a percentage (0-100).
        sentence_results: Detailed results for each sentence.
        submitted_at: When the activity was submitted.
        difficulty: Difficulty level of the activity.
    """

    activity_id: str
    student_id: UUID
    score: int = Field(ge=0)
    total: int = Field(ge=1)
    percentage: float = Field(ge=0, le=100)
    sentence_results: list[SentenceResult]
    submitted_at: datetime
    difficulty: str

File: frontend/src/types/sentence-builder.ts
/**
 * Sentence Builder Activity type definitions for Dream LMS.
 * Story 27.13: Sentence Builder Activity (Duolingo-Style)
 *
 * These types support Duolingo-style sentence building activities where
 * students arrange jumbled words into correct sentence order.
 */

/**
 * Valid difficulty levels for sentence builder
 */
export const DIFFICULTY_LEVELS = ["easy", "medium", "hard"] as const
export type DifficultyLevel = (typeof DIFFICULTY_LEVELS)[number]

/**
 * Difficulty level word count ranges
 */
export const DIFFICULTY_WORD_RANGES: Record<
  DifficultyLevel,
  { min: number; max: number }
> = {
  easy: { min: 4, max: 6 },
  medium: { min: 7, max: 10 },
  hard: { min: 11, max: 20 },
}

/**
 * Valid sentence count options
 */
export const SENTENCE_COUNT_OPTIONS = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] as const
export type SentenceCount = (typeof SENTENCE_COUNT_OPTIONS)[number]

/**
 * Request payload for generating a sentence builder activity
 */
export interface SentenceBuilderRequest {
  book_id: number
  module_ids?: number[]
  sentence_count?: number
  difficulty?: DifficultyLevel
  include_audio?: boolean
}

/**
 * A single sentence item with shuffled word bank (full view)
 */
export interface SentenceBuilderItem {
  item_id: string
  correct_sentence: string
  words: string[]
  word_count: number
  audio_url: string | null
  source_module_id: number
  source_page: number | null
  difficulty: string
}

/**
 * A single sentence item without correct answer (student view)
 */
export interface SentenceBuilderItemPublic {
  item_id: string
  words: string[]
  word_count: number
  difficulty: string
}

/**
 * Full sentence builder activity with correct answers (teacher view)
 */
export interface SentenceBuilderActivity {
  activity_id: string
  book_id: number
  module_ids: number[]
  sentences: SentenceBuilderItem[]
  difficulty: string
  include_audio: boolean
  created_at: string
}

/**
 * Public sentence builder activity without correct answers (student view)
 */
export interface SentenceBuilderActivityPublic {
  activity_id: string
  book_id: number
  module_ids: number[]
  sentences: SentenceBuilderItemPublic[]
  difficulty: string
  include_audio: boolean
  created_at: string
  sentence_count: number
}

/**
 * Request payload for submitting sentence answers
 */
export interface SentenceBuilderSubmission {
  answers: Record<string, string[]>
}

/**
 * Result for a single sentence after submission
 */
export interface SentenceResult {
  item_id: string
  submitted_words: string[]
  correct_sentence: string
  is_correct: boolean
  audio_url: string | null
}

/**
 * Overall sentence builder result after submission
 */
export interface SentenceBuilderResult {
  activity_id: string
  student_id: string
  score: number
  total: number
  percentage: number
  sentence_results: SentenceResult[]
  submitted_at: string
  difficulty: string
}

/**
 * Activity state for the player component
 */
export type SentenceActivityState =
  | "loading"
  | "ready"
  | "in_progress"
  | "checking"
  | "correct"
  | "incorrect"
  | "completed"
  | "error"

/**
 * Current sentence state during activity
 */
export interface CurrentSentenceState {
  itemId: string
  placedWords: string[]
  availableWords: string[]
  isChecking: boolean
  isCorrect: boolean | null
  attempts: number
}

/**
 * Human-readable labels for difficulty levels
 */
export const DIFFICULTY_LABELS: Record<DifficultyLevel, string> = {
  easy: "Easy (4-6 words)",
  medium: "Medium (7-10 words)",
  hard: "Hard (11+ words)",
}

/**
 * Descriptions for difficulty levels
 */
export const DIFFICULTY_DESCRIPTIONS: Record<DifficultyLevel, string> = {
  easy: "Short sentences with simple structure - perfect for beginners",
  medium:
    "Medium-length sentences with some complexity - good for intermediate learners",
  hard: "Long sentences with complex structure - challenging for advanced learners",
}

/**
 * Type guard to check if activity has correct answers (teacher view)
 */
export function isFullActivity(
  activity: SentenceBuilderActivity | SentenceBuilderActivityPublic,
): activity is SentenceBuilderActivity {
  return (
    activity.sentences.length > 0 && "correct_sentence" in activity.sentences[0]
  )
}

/**
 * Type guard to check if activity is public view
 */
export function isPublicActivity(
  activity: SentenceBuilderActivity | SentenceBuilderActivityPublic,
): activity is SentenceBuilderActivityPublic {
  return "sentence_count" in activity
}

/**
 * Check if a sentence item has correct answer (full view)
 */
export function isFullSentenceItem(
  item: SentenceBuilderItem | SentenceBuilderItemPublic,
): item is SentenceBuilderItem {
  return "correct_sentence" in item
}

/**
 * Get the progress text for current sentence
 */
export function getProgressText(currentIndex: number, total: number): string {
  return `${currentIndex + 1} of ${total}`
}

/**
 * Check if the placed words match the correct sentence
 */
export function checkAnswer(
  placedWords: string[],
  correctSentence: string,
): boolean {
  return placedWords.join(" ") === correctSentence
}

/**
 * Get feedback message based on correctness
 */
export function getFeedbackMessage(
  isCorrect: boolean,
  attempts: number,
): string {
  if (isCorrect) {
    if (attempts === 1) {
      return "Perfect! First try!"
    }
    return "Correct! Well done!"
  }
  return "Not quite right. Try again!"
}
